### PR TITLE
Adds GraphOfConvexSets::GetSolutionPath() and supporting introspection methods.

### PR DESCRIPTION
**Pull request by RussTedrake**
... and use it in GcsTrajectoryOptimization.
Also adds introspection methods that curate the incoming and outgoing edges from a vertex.

+@wrangelvid for feature review, please.
cc @TobiaMarcucci 
(note that one test will fail until #19790 lands, but I'd like to start the review process here)

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19792)

### CODE REVIEW

**Review by RussTedrake — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19792)* status: 1 unresolved discussion, LGTM missing from assignee wrangelvid, needs platform reviewer assigned, needs at least two assigned reviewers

___
*[`planning/trajectory_optimization/gcs_trajectory_optimization.cc` line 742 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-N_QLolR5e14sVZiT0Jz:-N_QLolR5e14sVZiT0K-:b-fabsw4) ([raw file](https://github.com/RobotLocomotion/drake/blob/c568248bf86ee510a093899f0da01d450ebdc51f/planning/trajectory_optimization/gcs_trajectory_optimization.cc#L742)):*
> ```cpp
>       // If no candidate edges are found, backtrack to the previous node and
>       // continue the search.
>       path_vertex_ids.pop_back();
> ```

btw -- there was a bug here before.  it should have popped the path_edges, too.  it's fixed and tested in the updated version.

**Review by wrangelvid — COMMENTED** [review]
Looks good to me after we resolve if we go ahead and use ID's or references. :lgtm: 

> Reviewed 4 of 6 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-:-N_WEwr334CNjImIknKU:b-vscdgn)* status: 2 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers

___
*[`geometry/optimization/graph_of_convex_sets.h` line 555 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-N_WDnh00E_Z9KhppGPZ:-N_WDnh00E_Z9KhppGP_:bc7km6d) ([raw file](https://github.com/RobotLocomotion/drake/blob/41120271644538b118fd47b1bd9cf1ff6bea5b0e/geometry/optimization/graph_of_convex_sets.h#L555)):*
> ```objc
>   `target` can be found in the solution.
>   */
>   Path GetSolutionPath(VertexId source_id, VertexId target_id,
> ```

Why using IDs if we could use references our pointers?
___
*[`geometry/optimization/graph_of_convex_sets.h` line 572 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-N_WDz1H-A0tt92QgBCD:-N_WDz1H-A0tt92QgBCE:b-rozhks) ([raw file](https://github.com/RobotLocomotion/drake/blob/41120271644538b118fd47b1bd9cf1ff6bea5b0e/geometry/optimization/graph_of_convex_sets.h#L572)):*
> ```objc
>   */
>   solvers::MathematicalProgramResult SolveConvexRestriction(
>       const std::set<EdgeId>& active_edge_ids,
> ```

Oh I see, the IDs come from here. Do you think it makes sense to stick with references for public use?
___
*[`geometry/optimization/test/graph_of_convex_sets_test.cc` line 1114 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-N_WCK-KBQZ1f4EzUFuL:-N_WCK-KBQZ1f4EzUFuM:b-e2l33a) ([raw file](https://github.com/RobotLocomotion/drake/blob/41120271644538b118fd47b1bd9cf1ff6bea5b0e/geometry/optimization/test/graph_of_convex_sets_test.cc#L1114)):*
> ```cpp
>   }
> 
>   // Now we artificially change the binaries in the result to cause
> ```

Very cool! I have been wondering how to properly test the backtracking.
___
*[`planning/trajectory_optimization/gcs_trajectory_optimization.cc` line 742 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-N_QLolR5e14sVZiT0Jz:-N_WCimLC5zXsTKC0y3_:b6zrrkh) ([raw file](https://github.com/RobotLocomotion/drake/blob/c568248bf86ee510a093899f0da01d450ebdc51f/planning/trajectory_optimization/gcs_trajectory_optimization.cc#L742)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

btw -- there was a bug here before.  it should have popped the path_edges, too.  it's fixed and tested in the updated version.
</blockquote></details>

Good catch.

**Review by RussTedrake — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19792)* status: 2 unresolved discussions, needs platform reviewer assigned, needs at least two assigned reviewers

___
*[`geometry/optimization/graph_of_convex_sets.h` line 555 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-N_WDnh00E_Z9KhppGPZ:-N_cOy-B32m-fnIamxva:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/41120271644538b118fd47b1bd9cf1ff6bea5b0e/geometry/optimization/graph_of_convex_sets.h#L555)):*
<details><summary><i>Previously, wrangelvid (David von Wrangel) wrote…</i></summary><blockquote>

Why using IDs if we could use references our pointers?
</blockquote></details>

Done.
___
*[`geometry/optimization/graph_of_convex_sets.h` line 572 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-N_WDz1H-A0tt92QgBCD:-N_cOzFGDNkrVGSPXesk:bav3m7x) ([raw file](https://github.com/RobotLocomotion/drake/blob/41120271644538b118fd47b1bd9cf1ff6bea5b0e/geometry/optimization/graph_of_convex_sets.h#L572)):*
<details><summary><i>Previously, wrangelvid (David von Wrangel) wrote…</i></summary><blockquote>

Oh I see, the IDs come from here. Do you think it makes sense to stick with references for public use?
</blockquote></details>

Done. (after a long discussion on the cpp slack channel, I am going to removed Ids from the public API.  i've updated this PR to be consistent with that, and will send a follow-up for the rest).

**Review by wrangelvid — COMMENTED** [review]
> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19792)* status: needs platform reviewer assigned, needs at least two assigned reviewers

**Comment by RussTedrake** [comment]
@drake-jenkins-bot linux-focal-clang-bazel-experimental-everything-valgrind-memcheck please

**Review by RussTedrake — COMMENTED** [review]
I'm pretty sure I've introduced some sort of memory error here.  but i've thrown every memcheck I know how to use (i'm limited by being on my mac while traveling), and can't find it... :-(

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-:-N_eJTZV7SFpGjdSaF7o:bufdzk1)* status: needs platform reviewer assigned, needs at least two assigned reviewers

**Comment by jwnimmer-tri** [comment]
I posted a more detailed answer to slack, but to summarize here for the wider audience: I think the problem is iterating over `set<const Edge*>` now, which is non-deterministic.  Pointers will sort differently depending on how `malloc` ended up randomizing things.  Apparently it matters in what order we add the edge/vertex costs and constraints into the program?

**Review by RussTedrake — COMMENTED** [review]
Done.  Thank you!  I believe everything is in order now.
Could I request
+@jwnimmer-tri for platform review, please?  To double check me on this.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-:-N_fG4hs2badLhtVNuGW:b-d1228l)* status: LGTM missing from assignee jwnimmer-tri(platform)

**Review by RussTedrake — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19792)* status: 1 unresolved discussion, LGTM missing from assignee sherm1(platform)

___
*[`geometry/optimization/graph_of_convex_sets.h` line 268 at r5](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-N_j3xDY9xGEFdmpB-9z:-N_jc_yUCgWnZNVJ1tc1:b-ctdvso) ([raw file](https://github.com/RobotLocomotion/drake/blob/5f315a3ecc478aafa92c9057293b4ec8cdc35894/geometry/optimization/graph_of_convex_sets.h#L268)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW if mutating a Vertex is a rare and signficant event, consider `mutable_u(()` and `mutable_v()` to make calling code more terrifying to readers and to avoid accidental use of the mutable case.
</blockquote></details>

I definitely considered it.  But I think that if folks have mutable access to the Edge, then I think it's fair to make it easy/convenient (not scary) for them to get the Vertex.

**Review by sherm1 — COMMENTED** [review]
Platform :lgtm:

> Reviewed 1 of 6 files at r3, 3 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-:-N_jlFJFECYejkzizh2R:b8qx53o)* status: 2 unresolved discussions

___
*[`geometry/optimization/test/graph_of_convex_sets_test.cc` line 885 at r5](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-N_jjRSl9KlfdHYyvzPb:-N_jjRSl9KlfdHYyvzPc:b-5lao36) ([raw file](https://github.com/RobotLocomotion/drake/blob/5f315a3ecc478aafa92c9057293b4ec8cdc35894/geometry/optimization/test/graph_of_convex_sets_test.cc#L885)):*
> ```cpp
>   const auto path = g_.GetSolutionPath(*source_, *target_, result);
>   ASSERT_EQ(path.size(), 1);
>   EXPECT_EQ(path[0], e_on_);
> ```

BTW consider verifying that the "vertex not from this graph" and "tolerance is nonsense" errors work (though not part of the documented API)
___
*[`geometry/optimization/test/graph_of_convex_sets_test.cc` line 1065 at r5](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-N_ji9rq3u_1Wb6D9dOx:-N_ji9rq3u_1Wb6D9dOy:be2o2pb) ([raw file](https://github.com/RobotLocomotion/drake/blob/5f315a3ecc478aafa92c9057293b4ec8cdc35894/geometry/optimization/test/graph_of_convex_sets_test.cc#L1065)):*
> ```cpp
> }
> 
> // A simple shortest-path problem where the continuous variables do not effect
> ```

nit: effect -> affect

**Review by RussTedrake — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19792)* status: :shipit: complete! all discussions resolved, LGTM from assignees sherm1(platform),wrangelvid

___
*[`geometry/optimization/test/graph_of_convex_sets_test.cc` line 885 at r5](https://reviewable.io/reviews/RobotLocomotion/drake/19792#-N_jjRSl9KlfdHYyvzPb:-N_k00TZ6yoAzEZpOKCe:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/5f315a3ecc478aafa92c9057293b4ec8cdc35894/geometry/optimization/test/graph_of_convex_sets_test.cc#L885)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW consider verifying that the "vertex not from this graph" and "tolerance is nonsense" errors work (though not part of the documented API)
</blockquote></details>

Done.

**Review by sherm1 — COMMENTED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19792)* status: :shipit: complete! all discussions resolved, LGTM from assignees sherm1(platform),wrangelvid

File: bindings/pydrake/geometry/geometry_py_optimization.cc [hunk 1]
```diff
@@ -671,6 +679,9 @@ void DefineGeometryOptimization(py::module m) {
             py::arg("source"), py::arg("target"),
             py::arg("options") = GraphOfConvexSetsOptions(),
             cls_doc.SolveShortestPath.doc_by_reference)
+        .def("GetSolutionPath", &GraphOfConvexSets::GetSolutionPath,
+            py::arg("source"), py::arg("target"), py::arg("result"),
+            py::arg("tolerance") = 1e-3, cls_doc.GetSolutionPath.doc)
```
Comment (jwnimmer-tri) [inline bindings/pydrake/geometry/geometry_py_optimization.cc:684]: See #20047.  Maybe this missing `py_reference_internal`?

**Comment by RussTedrake** [comment]
*[`bindings/pydrake/geometry/geometry_py_optimization.cc` line 684 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/19792#gh-1303651412:-NcZsku7EDG9uiULONAs:bx3r6mm) ([raw file](https://github.com/RobotLocomotion/drake/blob/03ce32a2d6e54139d6cfadddadc1193792f57577/bindings/pydrake/geometry/geometry_py_optimization.cc#L684)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

See #20047.  Maybe this missing `py_reference_internal`?
</blockquote></details>

You're right.  I've confirmed that as a fix, and will PR it.  Thanks.

**Pull request merged**